src/nvim/ops.c

```diff
@@ -2637,12 +2637,59 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
    * special characters (newlines, etc.).
    */
   if (regname == '.') {
-    (void)stuff_inserted((dir == FORWARD ? (count == -1 ? 'o' : 'a') :
-                          (count == -1 ? 'O' : 'i')), count, FALSE);
-    /* Putting the text is done later, so can't really move the cursor to
-     * the next character.  Use "l" to simulate it. */
-    if ((flags & PUT_CURSEND) && gchar_cursor() != NUL)
-      stuffcharReadbuff('l');
+    bool non_linewise_vis = (VIsual_active && VIsual_mode != 'V');
+
+    // PUT_LINE has special handling below which means we use 'i' to start.
+    char command_start_char = non_linewise_vis ? 'c' :
+        (flags & PUT_LINE ? 'i' : (dir == FORWARD ? 'a' : 'i'));
+
+    // To avoid being the affect of 'autoindent' on linewise puts, we create a
+    // new line with `:put _`.
+    if (flags & PUT_LINE) { do_put('_', NULL, dir, 1, PUT_LINE); }
+
+    // If given a count when pasting linewise, we stuff the readbuf with the
+    // dot register 'count' times split by newlines.
+    if (flags & PUT_LINE) {
+        stuffcharReadbuff(command_start_char);
+        for (; count > 0; count--) {
+            (void)stuff_inserted(NUL, 1, count != 1);
+            if (count != 1) { stuffcharReadbuff('\n'); }
+        }
+    } else {
+        (void)stuff_inserted(command_start_char, count, false);
+    }
+    // Putting the text is done later, so can't move the cursor to the next
+    // character.  Simulate it with motion commands after the insert.
+    if (flags & PUT_CURSEND) {
+        if (flags & PUT_LINE) { stuffReadbuff((char_u *)"j0");
+        } else {
+            // Ensure that we don't ring the bell from attempting to move into
+            // the space after the current line.
+            // We can stuff the readbuffer with "l" if
+            // 1) 'virtualedit' is "all" or "onemore"
+            // 2) We are not at the end of the line
+            // 3) We are not  (one past the end of the line && on the last line)
+            // The last condition is to account for a visual paste over a
+            // selection one past the end of the line joining the current line
+            // with the one below.
+            colnr_T line_len = (colnr_T)STRLEN(ml_get(curwin->w_cursor.lnum));
```

Comment:
I completely forgot to think about multibyte lines -- I'll go check that out. Thanks.

I don't think we can just disable bells, as then the user doesn't get a bell when the `".` register contains a backspace in a position that `'backspace'` doesn't allow it. I'll add a test for this.